### PR TITLE
Alusine JavaScript2 - homework week3

### PR DESCRIPTION
**Pull request by alquinex**

### CODE REVIEW

File: Week3/homework/step2-1.js [hunk 1]
```diff
@@ -1,8 +1,7 @@
 'use strict';
 
 function foo(func) {
-  // What to do here?
-  // Replace this comment and the next line with your code
+  func();
```
Comment (remarcmij) [inline Week3/homework/step2-1.js]: Yep, that is what was to do here.

File: Week3/homework/step2-1.js [hunk 2]
```diff
@@ -1,8 +1,7 @@
...
   console.log(func);
```
Comment (remarcmij) [inline Week3/homework/step2-1.js]: This console.log was supposed to be deleted.

File: Week3/homework/step2-2.js [hunk 3]
```diff
@@ -3,18 +3,27 @@
 function threeFive(startIndex, stopIndex, threeCallback, fiveCallback) {
   const numbers = [];
 
-  // Replace this comment and the next line with your code
+  for (let i = startIndex; i <= stopIndex; i++) {
+    numbers.push(i);
+  }
+  for (let i = startIndex; i <= stopIndex; i++) {
```
Comment (remarcmij) [inline Week3/homework/step2-2.js]: Although your code passes the test, you were asked to iterate through the `numbers` array you created in line 6-7. Can you please correct this?

File: Week3/homework/step2-3.js [hunk 4]
```diff
@@ -2,38 +2,39 @@
...
 
 // Use a 'do...while' loop
 function repeatStringNumTimesWithDoWhile(str, num) {
-  // eslint-disable-next-line prefer-const
   let result = '';
-
-  // Replace this comment and the next line with your code
-  console.log(str, num, result);
-
+  if (num > 0) {
```
Comment (remarcmij) [inline Week3/homework/step2-3.js]: 👍 for including this test before entering the `do-while` loop.

File: Week3/homework/step2-4.js [hunk 5]
```diff
@@ -1,7 +1,9 @@
 'use strict';
 
 function Dog() {
-  // add your code here
+  this.name = 'Whitey';
+  this.color = 'green';
+  this.numLegs = 4;
 }
 
 const hound = new Dog();
```
Comment (remarcmij) [inline Week3/homework/step2-4.js]: 👍

File: Week3/homework/step2-5.js [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
-  // eslint-disable-next-line
   let product = 1;
+  for (let i = 0; i < arr.length; i++) {
+    for (let j = 0; j < arr[i].length; j++) {
+      product *= arr[i][j];
+    }
+  }
 
-  // Replace this comment and the next line with your code
   console.log(arr, product);
```
Comment (remarcmij) [inline Week3/homework/step2-5.js]: This console.log is unneeded and should not be part of a pull request.

File: Week3/homework/step2-5.js [hunk 7]
```diff
@@ -1,10 +1,13 @@
 'use strict';
 
 function multiplyAll(arr) {
-  // eslint-disable-next-line
   let product = 1;
+  for (let i = 0; i < arr.length; i++) {
+    for (let j = 0; j < arr[i].length; j++) {
+      product *= arr[i][j];
```
Comment (remarcmij) [inline Week3/homework/step2-5.js]: Nice use of the short-hand multiplication operator.

File: Week3/homework/step2-6.js [hunk 8]
```diff
@@ -4,13 +4,27 @@ const arr2d = [[1, 2], [3, 4], [5, 6]];
...
+    for (let j = 0; j < arr[i].length; j++) {
+      for (let k = 0; k < arr[i][j].length; k++) {
+        flatArray3d.push(arr[i][j][k]);
+      }
+    }
+  }
+  return flatArray3d;
 }
 
 console.log(flattenArray2d(arr2d)); // -> [1, 2, 3, 4, 5, 6]
```
Comment (remarcmij) [inline Week3/homework/step2-6.js]: These solutions are fit-for-purpose and meet the test requirements. Any idea how to deal with an array of any dimension?

File: Week3/homework/step2-7.js [hunk 9]
```diff
@@ -20,4 +20,12 @@ f2(y);
...
-// Add your explanation as a comment here
+// In the first function, the function is called by directly
+// passing the value of the variable as the argument. Changing
+// the argument inside the function does not affect the variable
+// passed from outside. This is known as Pass by Value.
+
+// On the other hand we have Pass by Reference in the second function.
+// This is where the function is directly called by passing reference or
+// address of the variable as the argument. Changing the argument inside the
+// function affects the variable passed from outside the function.
```
Comment (remarcmij) [inline Week3/homework/step2-7.js]: I'm completely satisfied with your explanation 👍 .

File: Week3/homework/step3-bonus.js [hunk 10]
```diff
@@ -3,8 +3,7 @@
 const values = ['a', 'b', 'c', 'd', 'a', 'e', 'f', 'c'];
 
 function makeUnique(arr) {
-  // Replace this comment and the next line with your code
-  console.log(arr);
+  return arr.filter((index1, index2, index3) => index3.indexOf(index1) === index2);
```
Comment (remarcmij) [inline Week3/homework/step3-bonus.js]: While the code passes the test, this line is difficult to grasp as the names `index1`, `index2` and `index3` do not convey anything useful. In fact, only one of them is truly an index. If you look up the documentation of `filter` (google 'mdn filter') you will find the argument names `element`, `index` and `array`. You don't really need the third argument as you can already access the array through the `arr` variable.

Your code becomes easier to understand if you pick appropriate names:

```js
return arr.filter((element, index) => arr.indexOf(element) === index);
```

File: Week3/homework/step3.js [hunk 11]
```diff
@@ -1,8 +1,7 @@
 'use strict';
 
 function createBase(base) {
-  // Replace this comment and the next line with your code
-  console.log(base);
+  return sum => base + sum;
 }
 
 const addSix = createBase(6);
```
Comment (remarcmij) [inline Week3/homework/step3.js]: 👍

**Comment by alquinex** [comment]
> Hi @alquinex, there is a minor fix I would like you to make in step2-2 before I can approve the pull request. Take note of the other comments. Apart from that, everything looks good.

Hi @remarcmij , thanks for the feedback and for your time. The changes have been made.

File: Week3/homework/step2-2.js (hunk 3 above)
Comment (alquinex) [reply Week3/homework/step2-2.js]: I am not sure, but I am actually using a for loop here to iterate through the 'numbers' array.

Comment (remarcmij) [reply Week3/homework/step2-2.js]: But you are not using the `numbers` array. The variable `numbers` is not referenced at all in your second loop. I would expect something like:

```js
for (let i = 0; i < numbers.length; i++) {
  const number = numbers[i];
  if (number % 3 === 0) {
    ...
```

File: Week3/homework/step3-bonus.js (hunk 10 above)
Comment (alquinex) [reply Week3/homework/step3-bonus.js]: My bad...It's done.
Thanks!!

File: Week3/homework/step2-2.js (hunk 3 above)
Comment (alquinex) [reply Week3/homework/step2-2.js]: My bad...It's done.
Thanks!!

File: Week3/homework/step2-2.js [hunk 12]
```diff
@@ -3,18 +3,28 @@
...
-  // Replace this comment and the next line with your code
+  for (let i = startIndex; i <= stopIndex; i++) {
+    numbers.push(i);
+  }
+  for (let i = 0; i < numbers.length; i++) {
+    const number = numbers[i];
+    if (number % 3 === 0) {
+      threeCallback(number);
+    }
+    if (i % 5 === 0) {
```
Comment (remarcmij) [inline Week3/homework/step2-2.js]: Hm, this doesn't look right...

**Pull request closed without merging**